src/briefcase/integrations/subprocess.py

```diff
@@ -235,25 +294,46 @@ def stream_output(self, label, popen_process):
         :param label: A description of the content being streamed; used for
             to provide context in logging messages.
         :param popen_process: a running Popen process with output to print
+        :param stop_func: a Callable that returns True when output streaming
+            should stop and the popen_process should be terminated.
         """
+        output_streamer = threading.Thread(
+            name=f"{label} output streamer",
+            target=self._stream_output_thread,
+            args=(popen_process,),
+        )
         try:
-            while True:
-                # readline should always return at least a newline (ie \n)
-                # UNLESS the underlying process is exiting/gone; then "" is returned
-                output_line = ensure_str(popen_process.stdout.readline())
-                if output_line:
-                    self.command.logger.info(output_line)
-                elif output_line == "":
-                    # a return code will be available once the process returns one to the OS.
-                    # by definition, that should mean the process has exited.
-                    return_code = popen_process.poll()
-                    # only return once all output has been read and the process has exited.
-                    if return_code is not None:
-                        self._log_return_code(return_code)
-                        return
-
+            output_streamer.start()
+            if stop_func:
+                while output_streamer.is_alive() and not stop_func():
+                    time.sleep(0.5)
+            else:
+                output_streamer.join()
         except KeyboardInterrupt:
+            pass  # allow CTRL+C to gracefully stop streaming
+        finally:
             self.cleanup(label, popen_process)
+            output_streamer.join()
```

Comment:
After initially deciding a final `join` on the thread wasn't necessary because `cleanup` literally ensures the process the thread is monitoring is killed, I changed my mind during testing.

In the case of the keyboard interrupt test, pytest wouldn't always be able to capture the output in `capsys` because it got flushed through `stdout` from the thread too late....meaning that `stream_output` was returning control before the thread was truly done....or in the least before its streams were flushed.

Additionally, in thinking about a real world case where this might be a problem.....imagine a process sends an immense amount of data to `stdout`. Then the output streamer thread is likely to end up with a backlog in the pipe to be read. If the process terminates at this point, that thread will continue reading that pipe and printing output to the screen. Therefore, it would probably be better to let the thread finish before we return control back upstream; otherwise, upstream code (being executed in the main thread) could then try to print messages that would be interleaved with the output streamer's printing.

OTOH, if that scenario were to happen, having all that output on the screen may or may not be desirable anyway. Therefore, we may want to consider also killing the thread after a bit the same way we do with the process. Python doesn't expose an API to kill threads but we could always pass a stop flag to the thread for it to monitor. Probably overkill....but putting it out there.

Reply:
I think your instinct is right that the `join` is worth it. It's an edge case, but if you've got a process dumping large volumes of logs, there's a good chance the reason is because you're trying to debug some odd crash behavior - so making sure that the stream is fully purged before control is returned will avoid confusion.

I see what you're saying about extra safeties around killing the thread, but I think we can put that in the "wait and see" category. In my testing, I haven't noticed a problem with thread termination, so I'm happy to land this as-is. If we start seeing reports of odd behavior, we can revisit that as a bugfix/improvement.